core/src/util/builders.cpp

```diff
@@ -59,22 +45,72 @@ void Builders::buildPolygon(const Polygon& _polygon, float _height, PolygonBuild
         }
     }
 
-    // call the tesselator
-    glm::vec3 normal(0.0, 0.0, 1.0);
+    _ctx.earcut(_polygon);
+
+    size_t sumPoints = 0;
+    for (auto& line : _polygon) {
+        sumPoints += line.size();
+    }
 
-    _ctx.numVertices += _ctx.earcut.vertices.size();
+    size_t sumVertices = 0;
+    _ctx.used.assign(sumPoints, 0);
+    for (auto i : _ctx.earcut.indices) {
+        if (_ctx.used[i] == 0) {
+            _ctx.used[i] = 1;
+            sumVertices++;
+        }
+    }
+
+    uint16_t vertexDataOffset = _ctx.numVertices;
+    _ctx.numVertices += sumVertices;
     _ctx.sizeHint(_ctx.numVertices);
 
-    for (auto& p : _ctx.earcut.vertices) {
-        glm::vec3 coord(p[0], p[1], _height);
+    size_t dst = 0;
+    size_t modified = 0;
 
-        if (_ctx.useTexCoords) {
-            glm::vec2 uv(mapValue(coord.x, min.x, max.x, 0., 1.),
-                         mapValue(coord.y, min.y, max.y, 1., 0.));
+    size_t ring = 0;
+    size_t offset = 0;
+
+    for (size_t src = 0; src < _ctx.used.size(); src++) {
+
+        if (src - offset >= _polygon[ring].size()) {
+            offset += _polygon[ring].size();
+            ring += 1;
+        }
+        const Line& points = _polygon[ring];
+
+        if (_ctx.used[src] > 0) {
```

Comment:
K, got it, but yes I agree with @blair1618, it needs some comments/documentation love.